### ISSUE TITLE
Compile with stable Rust

### ISSUE THREAD

**Issue by SergioBenitez** [SergioBenitez/Rocket#19]
The following features need to be stabilized before Rocket can compile on stable:

 - [x] proc_macro_hygiene (rust-lang/rust#54727)
 - [x] associated_consts (rust-lang/rust#29646)
 - [x] conservative_impl_trait (rust-lang/rust#34511)
 - [x] drop_types_in_const (rust-lang/rust#33156)
 - [x] more_io_inner_methods (rust-lang/rust#41519)
 - [x] pub_restricted (rust-lang/rust#32409)
 - [x] struct_field_attributes (rust-lang/rust#41681)
 - [x] use_extern_macros (rust-lang/rust#35896)
 - [x] fn_box (rust-lang/rust#28796)
 - [x] try_from (rust-lang/rust#33417)

The following dependencies rely on nightly features and must be updated before Rocket can compile on stable:

 - [x] `pear` (https://github.com/SergioBenitez/Pear/commit/333c9bf5e2c516c23b17c7482b410f1448c1b3af)

Update (Feb 07, 2017): Added list of features used by Rocket.
Update (Feb 28, 2017): Added `lookup_host` feature.
Update (Mar 21, 2017): `pub_restricted` was stabilized!
Update (Mar 30, 2017): Added `never_type` feature.
Update (Apr 16, 2017): Added `concat_idents` feature.
Update (May 19, 2017): Added `struct_field_attributes` and `more_io_inner_methods` features.
Update (May 19, 2017): `concat_idents` is no longer used.
Update (May 19, 2017): Added links to tracking issues.
Update (May 20, 2017): `type_ascription` is no longer used.
Update (Jun 19, 2017): `struct_field_attributes` was stabilized!
Update (Jun 24, 2017): Added `try_trait` feature.
Update (Jul 1, 2017): `more_io_inner_methods` was stabilized!
Update (Jul 9, 2017): `associated_consts` was stabilized!
Update (Sep 7, 2017): `lookup_host` is no longer used.
Update (Sep 14, 2017): `drop_types_in_const` was stabilized!
Update (Sep 14, 2017): Added `decl_macro` feature.
Update (Mar 26, 2018): `conservative_impl_trait`, `never_type`, and `i128_type` were stabilized!
Update (Apr 22, 2018): Added `fnbox` feature.
Update (Apr 26, 2018): `never_type` stabilization was reverted (rust-lang/rust#50121).
Update (May 5, 2018): Swapped `macro_reexport ` for `use_extern_macros`.
Update (Jul 29, 2018): Added `crate_visibility_modifier` and `try_from` features.
Update (Aug 5, 2018): `custom_derive` is no longer used.
Update (Aug 17, 2018): `use_extern_macros` was stabilized!
Update (Sep 3, 2018): `const_fn` is no longer used.
Update (Sep 26, 2018): Added `label_break_value` feature.
Update (Oct 9, 2018): Removed compiler plugin features (`custom_attribute`, `plugin`).
Update (Oct 9, 2018): Updated `proc_macro` features.
Update (Mar 9, 2019): `try_from` was stabilized!
Update (Apr 13, 2019): `fnbox` is no longer used.
Update (Jul 9, 2019): `never_type` is no longer used.
Update (Jul 19, 2019): `decl_macro` is no longer used.
Update (Sep 9, 2019): `specialization` is no longer used.
Update (Sep 10, 2019): `label_break_value` is no longer used.
Update (Sep 18, 2019): `try_trait` is no longer used.
Update (Sep 21, 2019): `crate_visibility_modifier` is no longer used.
Update (May 19, 2020): `proc_macro_hygiene` was stabilized!
Update (Jul 16, 2020): `proc_macro_diagnostics` is no longer used.
Update (Jul 16, 2020): `proc_macro_span` is no longer used.
Update (Jul 21, 2020): `pear` was updated to be stable-compatible.

**Comment by steveklabnik** [comment]
Looks like this primarily depends on https://github.com/rust-lang/rust/issues/38356

**Comment by rrichardson** [comment]
Syntex should get this onto stable much more quickly, since the macros > 1.1 effort seems to be just beginning to congeal. 

https://news.ycombinator.com/item?id=13245975

**Comment by steveklabnik** [comment]
> since the macros > 1.1 effort seems to be just beginning to congeal.

This needs more than macros 1.1, it needs 2.0. (If it only needed 1.1, then it could go stable in 1.15)

**Comment by rrichardson** [comment]
I intended  `> 1.1`  to indicate "after 1.1"  but that wasn't very clear.   
I've seen the procedural macros effect referenced as 1.2 as well as 2.0.    2.0 does make more sense, since it is a much broader scope, as I understand it.

**Comment by mgattozzi** [comment]
I've started working on getting `syntex` to work with this until macros are stabilized (which might be a while). I've got time over the holidays from work so I'll post any progress I make!

**Comment by SergioBenitez** [comment]
Rocket uses much more than just plugins from nightly. The set of features Rocket currently uses are:

```rust
#![feature(plugin)]
#![feature(custom_derive)]
#![feature(custom_attribute)]
#![feature(specialization)]
#![feature(conservative_impl_trait)]
#![feature(drop_types_in_const)]
#![feature(associated_consts)]
#![feature(const_fn)]
```

These uses are important and contribute to the pithiness of the API, so they will stick around until they're stabilized. Until then, I don't believe Rocket will run on stable.

With that being said, I believe that the ergonomics of `syntex` are very poor, and I don't want to recommend a subpar solution to users. I'm all for getting Rocket to work on stable as soon as possible, but I'm not willing to compromise ergonomics and usability to make that happen.

**Comment by beefsack** [comment]
> I'm all for getting Rocket to work on stable as soon as possible, but I'm not willing to compromise ergonomics and usability to make that happen.

I'd love to see it go stable as soon as possible, but I feel you've got the right values for the project.  Accessibility is arguably more important for web frameworks than it is for other frameworks and libraries.

**Comment by mgattozzi** [comment]
Just from trying to get it to work with syntex it's hard. I might try some more this week but you're right. It might just be worth it to wait till those features stabilize.

**Comment by tbu-** [comment]
The `lookup_host` feature can be replaced by the `ToSocketAddrs` implementation of `&str`.

**Comment by FractalMatt** [comment]
Perhaps you could use something like:
https://github.com/Kimundi/rustc-version-rs

In order to allow stable features to compile using stable, but features that require nightly or beta could be used only when compiling from that branch.

Currently using Iron myself and would love to switch but it's a high security project and for security reasons, I want to avoid Rust nightly.

**Comment by FractalMatt** [comment]
It's been almost a year since this post originally... seems the check boxes in the original post are accurate?

Seems to me like getting Rocket out of nightly would be a huge boost for Rust in general but especially for Rocket.. the Rocket team or Rust team should really focus on stabilizing these functions (or finding alternatives?).

Why so slow to get there? Have you run this by the Rust higher ups and pushed them a little bit?

**Comment by messense** [comment]
[never_type will be stabilized in Rust 1.26.0](https://github.com/rust-lang/rust/pull/47630).

**Comment by messense** [comment]
[impl Trait is scheduled to be stabilized in Rust 1.26.0 too](https://github.com/rust-lang/rust/pull/49255)

**Comment by est31** [comment]
[i128 just got stabilized](https://github.com/rust-lang/rust/pull/49101#issuecomment-376315854) will ship in Rust 1.27.0.

**Comment by maxdeviant** [comment]
@messense PR #624 should handle those changes.

**Comment by mentalbrew** [comment]
The Rust community is currently in a major push toward releasing the Rust 2018 Edition. Has anybody investigated whether enough features that Rocket depends on in nightly will be available in that edition that the Rocket community could target that release for stable? It seems like it would be a win-win for both efforts if it was remotely feasible.

**Comment by jebrosen** [comment]
I glanced at the [2018 feature status](https://rust-lang-nursery.github.io/edition-guide/2018/status.html) and most of the important ones for Rocket are not listed:

* const_fn
* decl_macro
* never_type
* proc_macro, and several related features such as proc_macro_non_items
* specialization, which can apparently be worked around, but might be problematic to drop for existing rocket users

That does not necessarily mean that these features will not be stabilized this year, just that they are not being targeted as part of the Edition. `async`/`await`, however, *is* listed, and says 2018 will be the minimum edition. Depending on what async solution is implemented, Rocket might target the 2018 edition to make use of those ergonomics.

That being said, editions are cross-crate compatible. Whether Rocket ends up targeting the 2015 or 2018 edition should not significantly affect projects that use it.

**Comment by est31** [comment]
Just noting, `proc_macro` is in the process of being stabilized: https://github.com/rust-lang/rust/pull/52081

**Comment by SergioBenitez** [comment]
@est31 Unfortunately, the stabilizing subset is extremely limited and insufficient for Rocket's needs.

**Comment by cauebs** [comment]
Now there's an easier way to keep track of these things: https://forge.rust-lang.org/state-of-rust.html

**Comment by bestouff** [comment]
Either way it'd be nice to have a "tentative ETA" for stable Rocket. Even if it's "don't expect less than 2 years" that could help a lot people who might want to select it but wonder about how long they have to suffer in nightly.

**Comment by FractalMatt** [comment]
@SergioBenitez You should really reach out to the Rust team and ask for them to prioritize these issues, tell them it's holding up release of Rocket. Or maybe someone can volunteer to help fix these issues if they'll help point you in the right direction?

I know that getting Rocket-like web servers was being listed as a high priority for getting Rust adoption and Rocket looks like an otherwise amazing one.

Team can be found here:
https://www.rust-lang.org/en-US/team.html

Have communicated with Brian in the past who seemed particularly good at responding and helping us out on a far less important issue:
https://github.com/brson

**Comment by 0xcaff** [comment]
FYI, if you don't want to wait to build web services in rust, check out [actix](https://actix.rs/). It works on stable rust.

**Comment by dataf3l** [comment]
@0xcaff After trying multiple web frameworks, I have to say really like Rocket.
Iron, unfortunately, didn't do it for me...

I'm grateful for the efforts made by the iron team, and I value open source contributor's time as much as the next guy, but for me, I think Rocket suits fits my needs better.

**Comment by ErichDonGubler** [comment]
`proc_macro` [has been split](https://github.com/rust-lang/rust/issues/38356#issuecomment-426025407) into several subtasks -- which one(s) does Rocket depend on?

**Comment by Xaeroxe** [comment]
The following workaround has been used by the Amethyst game engine for the absence of a never type

```rust
pub enum Never {}
```

This type can't be instantiated and thus fills the same purpose.

**Comment by incubus8** [comment]
Hi, any plan to make it compile-able with stable Rust 1.31, https://blog.rust-lang.org/2018/12/06/Rust-1.31-and-rust-2018.html

Would love to see stable and production ready 👍

**Comment by est31** [comment]
@incubus8 see this blog post for details on the plans: https://rocket.rs/v0.4/news/2018-12-08-version-0.4/

**Comment by incubus8** [comment]
gotcha, sorry for the fuzz 

> @incubus8 see this blog post for details on the plans: https://rocket.rs/v0.4/news/2018-12-08-version-0.4/

**Comment by macpp** [comment]
Just noting: on the latest nightly `uniform_paths` were stabilised, together with ability to export macro_rules from modules: 
```rust
mod foo {
    macro_rules! bar123 { () => {} }
    pub(crate) use bar123 as bar;
}
foo::bar!();
```
It seems that it may be possible to remove `decl_macro` feature ( when targeting 2018 edition)

**Comment by jebrosen** [comment]
> It seems that it may be possible to remove decl_macro feature ( when targeting 2018 edition)

I played with the new export functionality you mention; it's unfortunate that it only works when the application crate targets 2018. I think it would be acceptable to implement the change if 2018 is required for another reason (e.g. `async/await`). But I wouldn't want to break compatibility with 2015 edition just to avoid `decl_macro` unless it's the only feature blocking stable, and maybe not even then.

**Comment by Centril** [comment]
@jebrosen I would suggest avoiding long term reliance on `decl_macro`; I don't think the language team has any bandwidth to devote the substantial time that `decl_macro` requires to stabilize it.

**Comment by abonander** [comment]
[`std::convert::Infallible`](https://doc.rust-lang.org/beta/std/convert/enum.Infallible.html) is stable in the upcoming release, that should be able to substitute for most uses of the never-type.

**Comment by jhpratt** [comment]
^^ I can quickly go through and issue a PR if that's something that's wanted.

**Comment by jebrosen** [comment]
At the very least, a PR will make it obvious where it's currently used and whether `Infallible` can fully replace its uses. It may not be accepted right way, especially if it would break existing code that uses the real never type.

**Comment by jhpratt** [comment]
@jebrosen Is there a minimum version Rocket guarantees, or is it just the current version? Asking because the initial checkout and `cargo check` returns warnings for both `try_from` and `transpose_result`. I'll naturally remove those as well if desired.

**Comment by jebrosen** [comment]
@jhpratt Only the `Infallible` change in one PR (at the very least separate commits) will be easier to split up later if we want to.

**Comment by jimcinbrisbane** [comment]
reqwest is almost stabled It's currently on 0.9.17
https://github.com/seanmonstar/reqwest

**Comment by Henning-K** [comment]
> reqwest is almost stabled It's currently on 0.9.17
> https://github.com/seanmonstar/reqwest

And how, if I may ask, is that relevant here?

**Comment by sound2gd** [comment]
hello, any update on this?

**Comment by jhpratt** [comment]
@sound2gd Rocket itself should be able to compile on stable as of today's release. Codegen is what still needs nightly.

**Comment by mastfissh** [comment]
If we wanted to use rocket on stable and we use codegen right now, how hard do you think it would be to remove the codegen usage?

Presumably we could use something like cargo-expand to see the generated code, and then replace the macro invocation with the generated code?

**Comment by jebrosen** [comment]
IIRC Some of `rocket`'s regular dependencies in `master` still use nightly features (most/all fixable by updating to a more recent version) and `rocket` depends on `rocket_codegen. Therefore it would take some work to actually do so. Using Rocket without codegen is usually recommended against and loses most of Rocket's value-add, so it's not something I'm interested in investing time into in favor of working on async and other improvements for 0.5.

> Presumably we could use something like cargo-expand to see the generated code, and then replace the macro invocation with the generated code?

I have usually found this to be impractical for anything other than testing, because you wouldn't want to and can't use the expansion of certain macros like `format!()`. In principle it could probably be done, but would not be very nice to maintain.

**Comment by mastfissh** [comment]
To summarize - it's probably not useful, easy or very maintainable to use rocket without codegen. Makes sense.

Thanks for the clear explanation 👍

**Comment by xfix** [comment]
`never_type` will be stabilized 1.41.0, so you can change `std::convert::Infallible` into `!`

**Comment by nicoburns** [comment]
Is there any chance we could use [this trick](https://github.com/RustPython/RustPython/pull/1669/files) to remove the dependency on `proc_macro_hygiene`, and get Rocket compiling on stable?

This [reddit discussion](https://www.reddit.com/r/rust/comments/emwjs1/heres_a_pr_showing_how_easy_it_is_to_remove_a/) suggests that it's a lot simpler than proc-macro-hack.

**Comment by jebrosen** [comment]
I think we could use that trick and get halfway there (we will have some diagnostics-related features to work out), but if I'm reading it right it would limit each function from calling `routes!()` or `uri!()` once.

**Comment by Kampfkarren** [comment]
Why doesn't Rocket just use proc-macro-hack directly? Implementation difficulty?

**Comment by jhpratt** [comment]
@Kampfkarren Read the comment right before yours 🙂

**Comment by Kampfkarren** [comment]
That's in reference to the trick used by RustPython, no? Not proc-macro-hack?

**Comment by jhpratt** [comment]
Ah, you're right. Either way, diagnostics would definitely take a big hit.

**Comment by jebrosen** [comment]
The limitation I was thinking of was something else and/or outdated - I believe both `proc-macro-hack` and the "extracted" version of it would be usable. When it comes down to that time, I think we could use this approach while the relevant part of `proc_macro_hygiene` is unstable.

**Comment by talglobus** [comment]
Status update?

**Comment by SergioBenitez** [comment]
A [stabilization PR](https://github.com/rust-lang/rust/pull/68717) for the final nightly-only feature Rocket makes use is under review! Looks like we're within a few months of Rocket compiling on stable in earnest.

**Comment by jhpratt** [comment]
Also worth tracking rust-lang/rust#68716, which is necessary to get that part of proc macro hygiene stabilized.

**Comment by runiq** [comment]
The guide and examples pretty much all still use `![feature(decl_macro)]` even though the OP says it's no longer used since July 2019. Removing the `![feature(decl_macro)]` line results in failed compiles.

I feel like I'm missing something: Is the `decl_macro` feature still needed or not? :thinking:

**Comment by xfix** [comment]
@runiq `decl_macro` is needed for 0.4. Future versions of rocket (0.5+) won't need it.

**Comment by futurechimp** [comment]
Hi, I've been holding off on making a decision about which HTTP framework to use for a project, while quietly watching progress on this one. Are there any updates about getting Rocket to run on stable Rust?

**Comment by HadrienG2** [comment]
As far as I know, this is currently blocked on the following dependency chain of Rust PRs, which all look well on their way:

- https://github.com/rust-lang/rust/pull/68717
- ...which is blocked on https://github.com/rust-lang/rust/pull/68716
- ...which is blocked on https://github.com/rust-lang/rust/pull/69041

**Comment by jhpratt** [comment]
To the best of my knowledge that is correct. One has finished FCP, one is currently in it, and the third has yet to begin it.

**If all three get merged by June 4, it will land on stable in Rust 1.45 on July 16.** The async version of Rocket will then be able to compile on stable, though as far as I know there's no timeframe on a 0.5 release.

**Comment by fylux** [comment]
Merged https://github.com/rust-lang/rust/pull/69041!
Thus https://github.com/rust-lang/rust/pull/68716 is ready to merge.

**Comment by runiq** [comment]
> Merged [rust-lang/rust#69041](https://github.com/rust-lang/rust/pull/69041)!
> Thus [rust-lang/rust#68716](https://github.com/rust-lang/rust/pull/68716) is ready to merge.

Aaand https://github.com/rust-lang/rust/pull/68716 has been merged as well.

**Comment by fylux** [comment]
The last remaining feature has just been merged! Finally Rocket will be able to compile with stable Rust (1.45).

**Comment by dataf3l** [comment]
I'm so happy to read this!!!, I've been tracking this bug for 2 years waiting with hope. good job guys :)

**Comment by kjedamzik** [comment]
Congratulations 🍾🥂

**Comment by a3d0n** [comment]
Awesome news 😃

**Comment by Mcat12** [comment]
> Update (May 19, 2019): proc_macro_hygiene was stabilized!

We're on 2020 now :wink:

**Comment by zolamk** [comment]
So when will we actually be able to build rocket with rust stable?

**Comment by Xaeroxe** [comment]
July 16th seems to be the day for stable, and June 4th is the day for beta.

**Comment by jhpratt** [comment]
Note that those dates also only apply to the upcoming 0.5 release. 0.4 will continue to require nightly for the time being.

**Comment by dnrusakov** [comment]
Amazing news! Congrats!

**Comment by Aaron1011** [comment]
Note that Rocket depends on `pear`, which still uses nightly features:

* https://github.com/SergioBenitez/Pear/blob/a4c5b1db60f22de88b62852d1a0da5ac992fc776/lib/src/lib.rs#L1-L3
* https://github.com/SergioBenitez/Pear/blob/a4c5b1db60f22de88b62852d1a0da5ac992fc776/codegen/src/lib.rs#L1-L3

**Comment by rursprung** [comment]
this is great news! is there a timeline for 0.5? i've seen the milestone which also seems to contain other features, some of which are already older issues: https://github.com/SergioBenitez/Rocket/milestone/8
would it be possible to make the 0.5 as soon as it contains everything needed to compile on stable (or, right now on beta)? this would be a huge step forward and move the library in the right direction. being able to build on a stable rust release would reduce the efforts for projects trying to use this library.

**Comment by rayrrr** [comment]
Hi @SergioBenitez thanks for this project. Given that Rocket depends on Pear, would you mind describing the purpose of Pear a bit in its own README? Seems like some util traits etc. at first glance...but with a good description in the README, perhaps some other contributors can help figure out how to remove the need for nightly there too.

**Comment by est31** [comment]
@rayrrr note that @Aaron1011 's links point to an outdated version of pear. In the current master version of pear (which rocket doesn't use yet as it uses the crates.io version), nightly usage has been removed from the pear_codegen crate, as well as `core_intrinsics` from the pear crate. The only two remaining feature gates are for `proc_macro_hygiene`, removal of which doesn't cause any errors, and `specialization`, which is apparently needed for conflicting implementations of the `Slice` trait.

**Comment by ambaxter** [comment]
I created a PR to remove all feature gates, though doing so appears require some mild lifetime changes in &str parameters. 

As of today the beta compiles all code *except* for https://github.com/SergioBenitez/Pear/blob/master/codegen/src/parser.rs#L168 . https://doc.rust-lang.org/proc_macro/struct.Span.html#method.join is nightly only and in all other compilers returns None. 

https://github.com/SergioBenitez/Pear/pull/27

**Comment by magnet** [comment]
> July 16th seems to be the day for stable, and June 4th is the day for beta.

Is that planning still accurate?  Excited to see Rocket on stable :)

**Comment by jhpratt** [comment]
Yes, Rocket itself should be able to compile on stable in two days. I'm not sure of the dependencies are all able to be built on beta currently, though.

**Comment by jplatte** [comment]
Pear was the last using Nightly features AFAICT from the discussion above and the linked issues & PRs. The only thing that remains to do for Rocket to compile on the current beta / upcoming stable is Pear being updated.

**Comment by leo-lb** [comment]
Stable is out now apparently, so it should compile on 1.45.x and later

**Comment by rayrrr** [comment]
Not quite there yet...https://github.com/SergioBenitez/Pear/issues/23#issuecomment-659499392

**Comment by npmccallum** [comment]
> Not quite there yet...[SergioBenitez/Pear#23 (comment)](https://github.com/SergioBenitez/Pear/issues/23#issuecomment-659499392)

https://github.com/SergioBenitez/Pear/pull/27

**Comment by rayrrr** [comment]
That's great to see @npmccallum, thanks for the heads up. I just tried building a Rocket project with the new stable Rust 1.45.0 and the build still failed around this issue, that's what brought me here. Hopefully all that's needed at this point for this use case is a release tag for Pear with that merged work and then updated references to the new Pear version in this repo. Looking forward to that!

**Comment by SergioBenitez** [comment]
Thank you, everyone, sincerely. :heart:

**Comment by jhpratt** [comment]
Congratulations Sergio!

**Comment by est31** [comment]
![rocket](https://user-images.githubusercontent.com/8872119/88134008-ce74d780-cbe3-11ea-89af-170c70263062.gif)

**Comment by l4l** [comment]
That's one small step for a crate; one giant leap for a rustaceans

**Comment by admwrd** [comment]
Goodbye Issue #19 Hello Stable! 🍾🍾🍾

**Comment by tungel** [comment]
Big thank you and congrats!!!

I've been waiting for this eagerly over the last few days. Kept on refreshing my browser tabs. Finally!

**Comment by mrcz** [comment]
Congrats. Will there be a 0.4.x release for this or will we have to wait for 0.5?

**Comment by zmilan** [comment]
Congrats. This is awesome news. Great work!

**Comment by deemoding** [comment]
Congratulations! Don't forget to update [doc](https://rocket.rs/v0.4/guide/getting-started)

**Comment by jebrosen** [comment]
> Congrats. Will there be a 0.4.x release for this or will we have to wait for 0.5?

It will happen in 0.5. Several of the features used by 0.4 (e.g. `try_trait`, `specialization`) are part of public API and won't be removed.

**Comment by zzFluke** [comment]
> Congratulations! Don't forget to update [doc](https://rocket.rs/v0.4/guide/getting-started)

And front-page as well

**Comment by tyt2y3** [comment]
Congratulations!

**Comment by hilltse** [comment]
The [getting started](https://rocket.rs/v0.4/guide/getting-started/) page still haven't been updated yet. 🚀🚀🚀

**Comment by RalfJung** [comment]
The current latest stable Rocket is 0.4, which still needs nightly Rust. So those updates will likely have to wait for Rocket 0.5.

**Comment by hktr92** [comment]
i can't wait for v0.5 to start experimenting. i might write a v2 app at work in rust and rocket :)

**Comment by DanielPower** [comment]
Is there an estimated timeline for 0.5 to be released?

**Comment by igalic** [comment]
@DanielPower , as  soon as it is done, unless things planned for the 0.5 milestone turn out to be to hard to do now, then possibly sooner (by not including them)

**Comment by jcbritobr** [comment]
Why stable is not priority? No one wants to use rocket in production?

**Comment by kjedamzik** [comment]
Why patience is not priority? The nightly is running quite stable in production here 🙂

**Comment by jcbritobr** [comment]
> 
> 
> Why patience is not priority? The nightly is running quite stable in production here 🙂

this not makes sense.

**Comment by PAStheLoD** [comment]
An IT system doesn't become "production ready" just because it has the name stable, nor does it become non-prod ready just because some component has "nightly" in its version string somewhere.

Inquiring about "when something will be done" or "why it's not done already" in a free & open source project will almost always result in someone (in this instance it's me) responding with "when someone does it" or "patches welcome".

**Comment by fallion** [comment]
@kjedamzik i think for a lot of people running on nightly builds (even if they may be stable) isn’t ideal.

**Comment by jcbritobr** [comment]
This is only a hispster project XD. Just for fun. I was wrong.

**Comment by schrieveslaach** [comment]
@jcbritobr, in [the chat](https://chat.mozilla.org/#/room/#rocket:mozilla.org) people are talking about the next release that is coming soon. So if we are a little bit patient, everything is fine.

**Comment by viniciusd** [comment]
> This is only a hispster project XD. Just for fun. I was wrong.

Chill @jcbritobr, there is no need to be [passive-]aggressive here, dude.

I understand you might be confused by the naming of the toolchains (stable vs nightly). I acknowledge they are indeed misleading.

Feel free to check the Rust release process at https://forge.rust-lang.org/release/process.html

Nightly indeed comes from the master branch and it may introduce regressions (and [it sporadically does](https://github.com/rust-lang/rust/issues/62562)), but the nightly "instability" is not about this. Stability is about the APIs, not about the compiler introducing bugs or breaking someone's code.

If your application compiles in a specific nightly version, there should not be any problem just because it is nightly. As long as it compiles, in general, you should be safe.

— Oh, but if it runs on nightly, my build will not be reproducible and it will change every day, oh noes!

Yeah, `nightly` itself will change, but [it is completely possible to pin a specific nightly version](https://doc.rust-lang.org/edition-guide/rust-2018/rustup-for-managing-rust-versions.html), e.g., 2020-11-30's nightly version! There is no need to be afraid of your toolchain changing without you knowing.

Finally, I am aware of a multitude of projects running Rocket in production. Using nightly does not make a project unstable, nor not-production-ready nor hipster as you said.

Feel free to reach out via the chat or other means if you require further clarification :)

**Comment by jcbritobr** [comment]
> 
> 
> @jcbritobr, in [the chat](https://chat.mozilla.org/#/room/#rocket:mozilla.org) people are talking about the next release that is coming soon. So if we are a little bit patient, everything is fine.

Hello. I'm not being passive/agressive. Look up these answers above. I only think bugs and stability comes first of new features. Postpone rocket on stable rust doesn't seems to be an intelligent decision. Nightly rust isn't stable rust, but if you guys think so very well. For me, its a hispter atitude. Just like for fun. Nothing to really help or agreggate the project.

**Comment by blackghost1987** [comment]
@jcbritobr the latest version of Rocket is already using Stable Rust! This is done already, that's why this issue is Closed, you are knocking on open doors. We are just waiting for the next Rocket release (0.5), because this is a breaking change.
It will be done shortly, the release depends on a couple more stuff though, compiling on Stable is just ONE of the new features, the list of the in-progress things can be found here:
https://github.com/SergioBenitez/Rocket/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.5.0

Also there's another question about the expected time of the release here: https://github.com/SergioBenitez/Rocket/issues/1476
As you can see SergioBenitez commented in November that he expects an RC for 0.5 to be released by the end of the month. It's only the 1st of December, so I think we can all be a little more patient with this.

In the meantime you can use Rocket from the master branch with Stable Rust. Or maybe help with one of the ongoing issues in the Milestones section to make the release happen faster.

**Comment by montdidier** [comment]
> An IT system doesn't become "production ready" just because it has the name stable, nor does it become non-prod ready just because some component has "nightly" in its version string somewhere.
> 

I kind of understand your point, but the reality is that that point of view is not widely shared, or even considered. The nomenclature is no doubt intended to have semantic meaning, and packaging crew working on numerous linux and BSD distributions see it _exactly_ this way. For somebody trying to get projects based on rocket into packages destined for various distributions this is a significant hurdle. I've had this problem with Alpine Linux and with OpenBSD just this month.

**Comment by PAStheLoD** [comment]
> For somebody trying to get projects based on rocket into packages destined for various distributions this is a significant hurdle.

Could you give examples of those projects?

> I've had this problem with Alpine Linux and with OpenBSD just this month.

Are many people using those projects with OpenBSD? OpenBSD is a very security-focused project, it's conservative in its scope, prefers simple instead of fast and fancy. If the OpenBSD maintainers feel they don't want to take something in then they'll keep coming up with excuses anyway. (Packaging policies are "premade excuses", they are not a guarantee for inclusion.)

So, even if that view is not shared, then I think a simpler version of that is probably applies here, which is _even if something is "stable" it doesn't make it production ready_. ([Bugs ](https://github.com/MrMEEE/bumblebee-Old-and-abbandoned/issues/123) happen. Ensuring quality is a process, determining fitness for a purpose requires a deliberation by someone who understands the needs and requirements of that purpose. A simple version sticker will never be able to do that.)

That said I completely understand distros' problem with managing "nightly", as it means an endless stream of packages. (Thus busy work.) It takes up space, it bloats the repository metadata, etc, etc. The solution is probably for those projects that depend on Rocket to do the packaging themselves and offer a repo that users can add, or a setup/update script. (I have no idea how that works on OpenBSD though. For Alpine it's pretty straightforward to host a repo, put APKs there and for users to set it up via /etc/apk/repositories and /etc/apk/keys .)

**Comment by TeXitoi** [comment]
I think the main problem for OpenBSD is that the package manager provide the latest stable, but not the nightly rust. Thus, you'll have to compile the compiler yourself if you want nightly as the rust project do not distribute rust for OpenBSD.

**Comment by TeXitoi** [comment]
Note that, as a workaround, you can use a stable compiler with `RUSTC_BOOTSTRAP=1`, and I think you'll have a rustc with feature gates usable.

**Comment by est31** [comment]
@TeXitoi please don't do that outside of local testing. It will cause issues the moment the compiler is updated.

**Comment by montdidier** [comment]
> Could you give examples of those projects?

The best example in my opinion is bitwarden_rs. It is quite a popular project and seems to be increasingly so. 

> Are many people using those projects with OpenBSD? OpenBSD is a very security-focused project, it's conservative in its scope, prefers simple instead of fast and fancy. If the OpenBSD maintainers feel they don't want to take something in then they'll keep coming up with excuses anyway. (Packaging policies are "premade excuses", they are not a guarantee for inclusion.)

Since the packages don't yet exist officially or via contrib as far as I know, I don't have a clear signal how who is using them on those platforms. I personally believe there is latent demand for them. I myself am trying to package them because I want to use them. 

I admit I don't know if they will accept a package before I present it, and the reason they fail to be accepted I don't always agree  with but even for my own use I still prefer the formal packaging system, and in the past I have experienced maintaining it this way it may eventually get accepted. My personal experiences with OpenBSD have been more positive, I suspect, based on your view expressed here.

> So, even if that view is not shared, then I think a simpler version of that is probably applies here, which is _even if something is "stable" it doesn't make it production ready_. ([Bugs ](https://github.com/MrMEEE/bumblebee-Old-and-abbandoned/issues/123) happen. Ensuring quality is a process, determining fitness for a purpose requires a deliberation by someone who understands the needs and requirements of that purpose. A simple version sticker will never be able to do that.)

I can see how this is true but is it really useful or good advice or a packagist? The fact is that stable, nightly and packaging policies are all systems to try to distill very complex things down to abstractions about which you can try to make reasonable decisions about. You're not claiming that nightly is just as stable as stable. It probably isn't on the whole. For a packagist trying to make the best decisions to further the interests of their chosen distribution, and probably with only casual knowledge about he myriad things over which they have purview, they will resort to leaning on these abstractions and for the most part that is completely reasonable and understandable.

> That said I completely understand distros' problem with managing "nightly", as it means an endless stream of packages. (Thus busy work.) It takes up space, it bloats the repository metadata, etc, etc. 

Absolutely. All good reasons that many will just want to keep it simple and stick with stable. 

> The solution is probably for those projects that depend on Rocket to do the packaging themselves and offer a repo that users can add, or a setup/update script. (I have no idea how that works on OpenBSD though. For Alpine it's pretty straightforward to host a repo, put APKs there and for users to set it up via /etc/apk/repositories and /etc/apk/keys .)

Yes, and no doubt some will choose to do that, but many projects will not be able to cover all that ground and there are packagists willing to do some of that lifting. Sticking with stable immediately lowers the bar for them, from a complexity point of view and from a distribution policy point of view. Neither of the challenges are insurmountable or even significant, but it is friction none the less.

In summary, all I argue is that relying on stable has more significant upside for rocket and projects that depend on it than I think has been acknowledged so far in this thread. I personally think rocket has a bright future and I'm looking forward to using it more myself.

**Comment by montdidier** [comment]
> I think the main problem for OpenBSD is that the package manager provide the latest stable, but not the nightly rust. Thus, you'll have to compile the compiler yourself if you want nightly as the rust project do not distribute rust for OpenBSD.

Yes this is correct and the packaging system also only will formally support crates from crates.io not from git directly. Perhaps that happens in the future, but not currently. There are workarounds but nothing that invokes pride in ones work.

**Comment by TeXitoi** [comment]
@est31 no more than updating a nightly, no? Note that the build infrastructure of firefox just do that.

But I agree that it is a workaround, and quite dirty one.
